### PR TITLE
Preserve from_docblock in TypeCombiner

### PR DESCRIPTION
**Pull request by VincentLanglet**
Closes https://github.com/vimeo/psalm/issues/8932

### CODE REVIEW

**Comment by orklah** [comment]
Thanks for pushin that. Unfortunately, there was a reason why I didn't merged it yet.

I *think* this test: https://github.com/vimeo/psalm/pull/8999/files#diff-5ad90d98b0768c1fdb94a7e73224d8709dd256fec470744ea1785b61ab72b39fL1088

is supposed to check that calling a methode that takes only `A` refines the original type from `A|null` (from docblock) to `A` (not from docblock)

The fact that you needed to go from `RedundantCondition` to `RedundantConditionGivenDocblockType` seems to indicate this is not working anymore. That's why I paused and I planned to check some things:
- check if the test was really supposed to test what I think
- check if it was already changed recently (to go from `RedundantConditionGivenDocblockType` to `RedundantCondition`)
- check if the thing that restrict the parameter after a method call did consider removing the from_docblock property

In order to know if it's a regression or just something that was not implemented and the test was changed when the from_docblock preservation broke

If you have time to look at this, I'd be glad :) Otherwise, I'll try to check this next week

File: tests/TypeReconciliation/RedundantConditionTest.php [hunk 1]
```diff
@@ -1546,6 +1546,27 @@ function f(array $p) : void {
                     }',
                 'error_message' => 'RedundantCondition',
             ],
+            'https://github.com/vimeo/psalm/issues/8932' => [
```
Comment (orklah) [inline tests/TypeReconciliation/RedundantConditionTest.php:1549]: Having a reference is a great idea but I still would like to have some kind of title Maybe a comment in the code itself? If someday, we were to migrate from github, we could loose the meaning behind the tests completely

**Comment by VincentLanglet** [comment]
> I _think_ this test: https://github.com/vimeo/psalm/pull/8999/files#diff-5ad90d98b0768c1fdb94a7e73224d8709dd256fec470744ea1785b61ab72b39fL1088
> 
> is supposed to check that calling a methode that takes only `A` refines the original type from `A|null` (from docblock) to `A` (not from docblock)
> 
> The fact that you needed to go from `RedundantCondition` to `RedundantConditionGivenDocblockType` seems to indicate this is not working anymore.

Indeed you're right.

> * check if the test was really supposed to test what I think

When reading again the test, it seems so.

> * check if it was already changed recently (to go from `RedundantConditionGivenDocblockType` to `RedundantCondition`)

No, this wasn't changed recently.

> * check if the thing that restrict the parameter after a method call did consider removing the from_docblock property

Seems like it need to be improve

**Comment by VincentLanglet** [comment]
> If you have time to look at this, I'd be glad :) Otherwise, I'll try to check this next week

To me, it's an issue when creating the TArray type https://github.com/vimeo/psalm/issues/8932#issuecomment-1364664519

**Comment by VincentLanglet** [comment]
The issue seems to be inside the
```
TypeCombiner::combine
```
method.

The input is types:
```
[Array(from_docblock = true), Null(from_docblock = true)]
```
but when `getArrayTypeFromGenericParams` is called there is 
```
$array_type = new TArray($generic_type_params);
```
which doesn't care about the previous `from_docblock` part.

I tried adding `from_docblock` everywhere in the TypeCombiner and it seems to solve everything.
I dunno if I missed some or if we have to remove some.

**Comment by orklah** [comment]
Seems like CI is failing, are you stuck?

**Comment by VincentLanglet** [comment]
> Seems like CI is failing, are you stuck?

Didn't saw the last commit introduced an error. We could have wait for a long time ^^'.
This is fixed now @orklah.

**Comment by orklah** [comment]
Thanks!

**Pull request merged**